src/Functions/fromReadable.h

```diff
@@ -0,0 +1,93 @@
+#include <base/types.h>
+#include <boost/algorithm/string/case_conv.hpp>
+
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/ReadHelpers.h>
+#include <string_view>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int TOO_FEW_ARGUMENTS_FOR_FUNCTION;
+    extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int BAD_ARGUMENTS;
+    extern const int UNEXPECTED_DATA_AFTER_PARSED_VALUE;
+    extern const int CANNOT_PARSE_INPUT_ASSERTION_FAILED;
+}
+
+template <typename Impl>
+class FunctionFromReadable : public IFunction
+{
+public:
+    static constexpr auto name = Impl::name;
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<FunctionFromReadable<Impl>>(); }
+
+    String getName() const override { return name; }
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+    size_t getNumberOfArguments() const override { return 1; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args
+        {
+            {"readable_size", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isString), nullptr, "String"},
+        };
+        validateFunctionArgumentTypes(*this, arguments, args);
+
+        return std::make_shared<DataTypeFloat64>();
+    }
+    
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        auto col_to = ColumnFloat64::create();
+        auto & res_data = col_to->getData();
+
+        for (size_t i = 0; i < input_rows_count; ++i)
+        {   
+            std::string_view str = arguments[0].column->getDataAt(i).toView();
+            ReadBufferFromString buf(str);
+            // tryReadFloatText does seem to not raise any error when there is leading whitespace so we check it explicitly
+            skipWhitespaceIfAny(buf);
+            if (buf.getPosition() > 0)
+                throw_exception(ErrorCodes::CANNOT_PARSE_INPUT_ASSERTION_FAILED, "Leading whitespace is not allowed", str);
```

Comment:
(Side note:  Here is where the normal function, `*OrNull` and `*OrZero` would differ. To implement these variants, the code in l. 52-76 could be moved into a separate function with the error handling as a template parameter (e.g. of enum class type, see `enum class ErrorHandling` in src/Functions/parseDateTime.cpp for an example). We could then use a `constexpr if` here to handle the error in an appropriate way.)

Reply:
Should be fine for now.